### PR TITLE
Drag library upto date

### PR DESCRIPTION
**Pull request by reedy**

### CODE REVIEW

**Comment by reedy** [comment]
Phan isn't "voting" yet, but passes:

```
reedy@ubuntu64-web-esxi:~/hamcrest-html-matchers$ composer phan
> phan -d . --long-progress-bar
[info] Disabling Xdebug: Phan is around five times as slow when Xdebug is enabled (Xdebug only makes sense when debugging Phan itself)
[info] To run Phan with Xdebug, set the environment variable PHAN_ALLOW_XDEBUG to 1.
[info] To disable this warning, set the environment variable PHAN_DISABLE_XDEBUG_WARN to 1.
[info] To include function signatures of Xdebug, see .phan/internal_stubs/xdebug.phan_php
[debug] Checking PHAN_ALLOW_XDEBUG
[debug] Because xdebug was installed, Phan will restart.
[debug] The Xdebug extension is loaded (2.9.6)
[debug] Async signals enabled (SIGINT = SIG_IGN)
[debug] Process restarting (PHAN_ALLOW_XDEBUG=internal|2.9.6|1|*|*)
[debug] Running '/usr/bin/php7.4' '-n' '-c' '/tmp/kIYKWU' '/home/reedy/hamcrest-html-matchers/vendor/bin/phan' '-d' '.' '--long-progress-bar'
Parsing files...
░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░   54 / 1089 (  5%) 41MB
░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░  108 / 1089 ( 16%) 60MB
░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░  162 / 1089 ( 16%) 60MB
░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░  216 / 1089 ( 33%) 75MB
░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░  270 / 1089 ( 33%) 75MB
░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░  324 / 1089 ( 33%) 75MB
░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░  378 / 1089 ( 57%) 92MB
░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░  432 / 1089 ( 57%) 92MB
░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░  486 / 1089 ( 57%) 92MB
░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░  540 / 1089 ( 57%) 92MB
░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░  594 / 1089 ( 57%) 92MB
░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░  648 / 1089 ( 73%) 108MB
░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░  702 / 1089 ( 73%) 108MB
░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░  756 / 1089 ( 73%) 108MB
░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░  810 / 1089 ( 84%) 131MB
░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░  864 / 1089 ( 84%) 131MB
░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░  918 / 1089 ( 84%) 131MB
░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░  972 / 1089 ( 95%) 152MB
░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░ 1026 / 1089 ( 95%) 152MB
░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░ 1080 / 1089 (100%) 165MB
░░░░░░░░░                                              1089 / 1089 (100%) 165MB
Analyzing classes...
░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░ 185MB
Analyzing functions...
░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░ 186MB
Analyzing methods...
░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░ 189MB
Analyzing files...
░░░░░░░░░░░░                                           12 / 12 (100%) 195MB

[debug] Restarted process exited 0
reedy@ubuntu64-web-esxi:~/hamcrest-html-matchers$ 
```

**Review by jdforrester — APPROVED** [review]

**Comment by reedy** [comment]
PHP 8.0 failing to install looks travis specific.

7.3 / `--prefer-lowest` is failing

Is there any value in `--prefer-lowest`, really?

It just looks to cause weirdness

```
> phpunit

PHPUnit 8.5.0 by Sebastian Bergmann and contributors.

Runtime:       PHP 7.3.27 with Xdebug 3.0.2

Configuration: /home/travis/build/wmde/hamcrest-html-matchers/phpunit.xml.dist

Code coverage needs to be enabled in php.ini by setting 'xdebug.mode' to 'coverage'

Script phpunit handling the phpunit event returned with error code 2

Script composer phpunit handling the test event returned with error code 2

Script composer test handling the ci event returned with error code 2

The command "composer ci" exited with 2.
cache.2

store build cache
```

**Comment by reedy** [comment]
Let's drop PHP 8 testing for now..

**Comment by reedy** [comment]
Tests pass fine on 7.2, 7.3, 7.4...

**Comment by reedy** [comment]
Will need GitHub actions enabling here I think...

**Comment by jdforrester** [comment]
Happy to help move this to the Wikimedia org with our unlimited GHA account if that's useful.

**Comment by reedy** [comment]
Minor tweaks needed, but tests now passing on PHP 8.0 too... https://github.com/reedy/hamcrest-html-matchers/actions/runs/671952382

**Comment by manicki** [comment]
> Happy to help move this to the Wikimedia org with our unlimited GHA account if that's useful.

WMDE github org has no limit on github actions either

**Comment by jdforrester** [comment]
> > Happy to help move this to the Wikimedia org with our unlimited GHA account if that's useful.
> 
> WMDE github org has no limit on github actions either

Oh, excellent, never mind then. :-)

File: tests/ComplexTagMatcherTest.php [hunk 1]
```diff
@@ -2,135 +2,108 @@
...
-	 */
-	public function canNotCreateMatcherWithChildElement() {
-		$this->setExpectedException( \Exception::class );
+	public function testCanNotCreateMatcherWithChildElement() {
+		$this->expectException( Exception::class );
 		ComplexTagMatcher::tagMatchingOutline( '<p><b></b></p>' );
 	}
 
-	/**
-	 * @test
```
Comment (Ladsgroup) [inline tests/ComplexTagMatcherTest.php:57]: ugh

File: tests/XmlNodeRecursiveIteratorTest.php [hunk 2]
```diff
@@ -42,15 +45,14 @@ private function createDomNodeListFromHtml( $html ) {
 	}
 
 	/**
-	 * @param $recursiveIterator
+	 * @param Traversable $recursiveIterator
```
Comment (Ladsgroup) [inline tests/XmlNodeRecursiveIteratorTest.php:48]: This could be a typehint but meh.

**Review by Ladsgroup — APPROVED** [review]

File: src/DirectChildElementMatcher.php [hunk 3]
```diff
@@ -54,6 +58,7 @@ protected function matchesSafelyWithDiagnosticDescription( $item, Description $m
 			return $directChildren->length !== 0;
 		}
 
+		$child = null;
```
Comment (thiemowmde) [inline src/DirectChildElementMatcher.php:61]: This line is technically dead code.

**Review by thiemowmde — COMMENTED** [review]
Welcome to the world of Aleksey. Good to see stuff like these `@test` been gone. Thanks a lot for taking care of this library!

**Comment by manicki** [comment]
Thanks for the updates indeed!

> Welcome to the world of Aleksey.

No need to get personal, though.

File: src/DirectChildElementMatcher.php (hunk 3 above)
Comment (reedy) [reply src/DirectChildElementMatcher.php:61]: Yeah, it felt cleaner than adding some sort of phan/other type suppression.

**Pull request merged**